File: sort/heap_sort/java/HeapSort.java
import java.util.*;
public class HeapSort
{
	public void sort(int arr[])
	{
		int n = arr.length;
		for (int i = n / 2 - 1; i >= 0; i--)
			heap(arr, n, i);
		for (int i=n-1; i>=0; i--)
		{
			int temp = arr[0];
			arr[0] = arr[i];
			arr[i] = temp;
			heap(arr, i, 0);
		}
	}
	void heap(int arr[], int n, int i)
	{
		int largest = i;
		int l = 2*i + 1;
		int r = 2*i + 2;
		if (l < n && arr[l] > arr[largest])
			largest = l;
		if (r < n && arr[r] > arr[largest])
			largest = r;
		if (largest != i)
		{
			int swap = arr[i];
			arr[i] = arr[largest];
			arr[largest] = swap;
			heap(arr, n, largest);
		}
	}
	static void printArray(int arr[])
	{
		int n = arr.length;
		for (int i=0; i<n; ++i)
			System.out.print(arr[i]+" ");
		System.out.println();
	}
	public static void main(String args[])
	{
		Scanner s=new Scanner(System.in);
		int arr[] = new int[5];
		int n = arr.length;
		System.out.println("enter array elements");
		for(int i=0;i<5;i++)
		arr[i]=s.nextInt();
		HeapSort ob = new HeapSort();
		ob.sort(arr);

		System.out.println("Sorted array is");
		printArray(arr);
	}
}

File: sort/heap_sort/golang/heapsort.go
package main

import (
	"fmt"
	"math"
)

func parent(node int) int {
	return int(math.Floor((float64(node) - 1.0) / 2.0))
}

func siftDown(a []int, start, end int) {
	root := start

	for 2*root+1 <= end {
		child := 2*root + 1
		swap := root

		if a[swap] < a[child] {
			swap = child
		}

		if child+1 <= end && a[swap] < a[child+1] {
			swap = child + 1
		}

		if swap == root {
			break
		} else {
			a[root], a[swap] = a[swap], a[root]
			root = swap
		}
	}
}

func heapify(a []int, count int) {
	start := parent(count - 1)

	for start >= 0 {
		siftDown(a, start, count-1)
		start = start - 1
	}
}

// HeapSort sorts an array using the heap sort algorithm
func HeapSort(a []int) []int {
	n := len(a)

	heapify(a, n)

	for end := n - 1; end > 0; {
		a[end], a[0] = a[0], a[end]
		end = end - 1
		siftDown(a, 0, end)
	}

	return a
}

func main() {
	a := []int{-1, 4, 9, 0, -10, 5, 894, 3493, -54958, 4}
	fmt.Println(a)
	fmt.Println("Sorted array: ", HeapSort(a))
}
